test/resolve-reference.js

```diff
@@ -0,0 +1,175 @@
+    common.resolveReference('💥', (err) => {
+      t.type(err, Error)
+      t.equal(err.message, 'Invalid resource reference "💥"')
+      t.end()
+    })
+  })
+
+  t.test('should return an error when the reference has an unknown resource type', (t) => {
+    common.resolveReference('Nope/1', (err) => {
+      t.type(err, Error)
+      t.equal(err.message, 'Invalid resource reference "Nope/1"')
+      t.end()
+    })
+  })
+
+  t.test('should return an error when the reference has an invalid id', (t) => {
+    common.resolveReference('Patient/#1', (err) => {
+      t.type(err, Error)
+      t.equal(err.message, 'Invalid resource reference "Patient/#1"')
+      t.end()
+    })
+  })
+
+  t.test('should return an error when the reference is an internal fragment', (t) => {
+    common.resolveReference('#1', (err) => {
+      t.type(err, Error)
+      t.equal(err.message, 'Invalid resource reference "#1"')
+      t.end()
+    })
+  })
+
+  t.test('should return null when the referenced resource does not exist', (t) => {
+    common.resolveReference('Patient/d959d05f-9e57-4897-be5a-362906e04558', (err, resolvedResource) => {
+      t.error(err)
+      t.equal(resolvedResource, null)
+      t.end()
+    })
+  })
+
+  t.test('should return the correct resource when resolving relative references', withDB((t, db) => {
+    createTestPatient(db, (err, referencedPatient) => {
+      if (err) {
+        return t.threw(err)
+      }
+      common.resolveReference(
+        `Patient/${referencedPatient.id}`,
+        (err, resolvedResource) => {
+          t.error(err)
+          t.deepEqual(resolvedResource, referencedPatient)
+          t.end()
+        }
+      )
+    })
+  }))
+
+  t.test('should return the correct resource when resolving absolute references', withDB((t, db) => {
+    createTestPatient(db, (err, referencedPatient) => {
+      if (err) {
+        return t.threw(err)
+      }
+      common.resolveReference(
+        `http://localhost/Patient/${referencedPatient.id}`,
+        (err, resolvedResource) => {
+          t.error(err)
+          t.deepEqual(resolvedResource, referencedPatient)
+          t.end()
+        }
+      )
+    })
+  }))
+
+  t.test('should return the correct resource when resolving relative references with a version', withDB((t, db) => {
+    createTestPatientHistory(db, (err, referencedPatient) => {
+      if (err) {
+        return t.threw(err)
+      }
+      common.resolveReference(
+        `Patient/${referencedPatient.id}/_history/${referencedPatient.meta.versionId}`,
+        (err, resolvedResource) => {
+          t.error(err)
+          t.deepEqual(resolvedResource, referencedPatient)
+          t.end()
+        }
+      )
+    })
+  }))
+
+  t.end()
+
+  function withDB (test) {
+    return (t) => {
+      mongo.getDB((err, db) => {
+        if (err) {
+          return t.threw(err)
+        }
+        test(t, db)
+      })
+    }
+  }
+
+  function createTestPatient (db, callback) {
```

Comment:
@bausmeier qq: is there a reason why you wrote this method instead of using the one in the test-env/init.js file?

Reply:
@nthfloor2 I didn't know that the others method existed. They also seem much more complicated than I need for this test and I don't like the idea of tests sharing fixture setup since it means that making changes to one test can break another.

Reply:
@bausmeier ok cool, makes sense, those methods were using full api requests, which is good for integration tests but a bit overkill for this.